### PR TITLE
Support boot image from container

### PR DESCRIPTION
**Pull request by stalamudupula**
Add support to load the boot image from container.
Container must be signed using the same private key
as the key used to sign IAS (i.e. IAS_PRIVATE_KEY).
ENABLE_CONT_BOOT needs to be set to 1 in BoardConfig.py
to enable boot from container.

Signed-off-by: Sai Talamudupula <sai.kiran.talamudupula@intel.com>

### CODE REVIEW

File: PayloadPkg/OsLoader/OsLoader.c [hunk 1]
```diff
@@ -154,8 +143,126 @@ ParseIasImage (
...
+  )
+{
+  EFI_STATUS                  Status;
+  CONTAINER_HDR              *ContainerHdr;
+  UINT64                      ComponentName;
+  LOADER_COMPRESSED_HEADER   *LzHdr;
+  IMAGE_DATA                  File[MAX_IAS_SUB_IMAGE];
+  UINT8                       Index;
+
+  ContainerHdr = (CONTAINER_HDR  *)LoadedImage->IasImage.Addr;
```
Comment (mauricema) [inline PayloadPkg/OsLoader/OsLoader.c]: Should we check the container signature here ?    Only handle signature with 'BOOT' ?

File: PayloadPkg/OsLoader/OsLoader.h [hunk 2]
```diff
@@ -74,6 +75,7 @@
 #define LOADED_IMAGE_LINUX       BIT2
 #define LOADED_IMAGE_PE32        BIT3
 #define LOADED_IMAGE_FV          BIT4
+#define LOADED_IMAGE_CONT        BIT5
```
Comment (mauricema) [inline PayloadPkg/OsLoader/OsLoader.h]: CONT usually means Continue.   Let us use full name CONTAINER

File: PayloadPkg/OsLoader/OsLoader.c [hunk 3]
```diff
@@ -11,79 +11,68 @@ UINT32  *mPreOsCheckerEntry;
...
-  AddMeasurePoint (0x4080);
-
-  ZeroMem (File, sizeof (File));
-  NumFiles = IasGetFiles (IasImage, sizeof (File) / sizeof ((File)[0]), File);
-  DEBUG ((DEBUG_INFO, "IAS size = 0x%x, file number: %d\n", LoadedImage->IasImage.Size, NumFiles));
+  Status = EFI_SUCCESS;
 
   if (NumFiles == 1) {
-    // This is not valid IAS use case, at least 2 files in IAS image.
+    // This is not valid IAS/PODS use case, at least 2 files in image.
```
Comment (mauricema) [inline PayloadPkg/OsLoader/OsLoader.c]: Let us use CONTAINER instead IAS/PODS.   IAS will be deprecated.

File: PayloadPkg/OsLoader/OsLoader.c [hunk 4]
```diff
@@ -11,79 +11,68 @@ UINT32  *mPreOsCheckerEntry;
...
 
   @param[in]      BootOption    Current boot option
+  @param[in]      File          Array of individual Files
   @param[in, out] LoadedImage   Loaded Image information.
+  @param[in]      ImageType     {NORMAL, CLASSIC, MULTIBOOT}
 
-  @retval  RETURN_SUCCESS       Parse IAS image successfully
-  @retval  Others               There is error when parsing IAS image.
+  @retval  RETURN_SUCCESS       Update LoadedImage successfully
+  @retval  Others               There is error updating.`
```
Comment (mauricema) [inline PayloadPkg/OsLoader/OsLoader.c]: why have an extra char at the end ?

File: BuildLoader.py [hunk 5]
```diff
@@ -190,6 +190,7 @@ def __init__(self, *args, **kwargs):
 		self.ENABLE_GRUB_CONFIG    = 0
 		self.ENABLE_SMBIOS         = 0
 		self.ENABLE_LINUX_PAYLOAD  = 0
+		self.ENABLE_CONT_BOOT	   = 0
```
Comment (mauricema) [inline BuildLoader.py]: Use full CONTAINER instead.

File: PayloadPkg/OsLoader/OsLoader.c (hunk 3 above)
Comment (mauricema) [reply PayloadPkg/OsLoader/OsLoader.c]: Or just say  "image" to be more generic.

File: PayloadPkg/OsLoader/OsLoader.c [hunk 6]
```diff
@@ -154,8 +143,126 @@ ParseIasImage (
...
+            File[Index].Addr = (UINT8 *)LzHdr + sizeof(LOADER_COMPRESSED_HEADER);
+            File[Index].Size = LzHdr->Size;
+          }
+        }
+      } else {
+        // Use Load to decompress to a new buffer
+        Status = LoadComponent (ContainerHdr->Signature, (UINT32) ComponentName, (VOID **)&File[Index].Addr, &File[Index].Size);
+      }
+    }
+    Index++;
```
Comment (mauricema) [inline PayloadPkg/OsLoader/OsLoader.c]: Pls consider error condition.  Index should only be increased Status is SUCCESS.

File: BootloaderCommonPkg/Include/Library/IasImageLib.h [hunk 7]
```diff
@@ -12,6 +12,7 @@
 #define IAS_MAGIC_PATTERN   0x2E6B7069 // ".kpi"
 #define MAX_IAS_SUB_IMAGE   32
 #define MAX_IAS_IMAGE_SIZE  0x19000000
+#define CONT_MAGIC_PATTERN  0x544F4F42 // "BOOT"
```
Comment (mauricema) [inline BootloaderCommonPkg/Include/Library/IasImageLib.h]: BOOT_MAGIC_PATTERN

File: BootloaderCommonPkg/Include/Library/IasImageLib.h [hunk 8]
```diff
@@ -12,6 +12,7 @@
 #define IAS_MAGIC_PATTERN   0x2E6B7069 // ".kpi"
 #define MAX_IAS_SUB_IMAGE   32
 #define MAX_IAS_IMAGE_SIZE  0x19000000
+#define BOOT_MAGIC_PATTERN  0x544F4F42 // "BOOT"
```
Comment (myguitar) [inline BootloaderCommonPkg/Include/Library/IasImageLib.h]: Better to use SIGNATURE_32 ('B', 'O', 'O', 'T')?

File: BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c [hunk 9]
```diff
@@ -276,6 +276,26 @@ AuthenticateComponent (
...
+
+  @retval                       COMP_TYPE_PUBKEY_OS for CONTAINER boot image
+                                COMP_TYPE_PUBKEY_CFG_DATA, otherwise
+**/
+UINT8
+GetContainerKeyTypeBySig (
+  IN  UINT32    ContainerSig
+  )
+{
+  if (ContainerSig == SIGNATURE_32 ('B', 'O', 'O', 'T')) {
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c]: Can we use BOOT_MAGIC_PATTERN here? A single signature definition would be better.

File: PayloadPkg/OsLoader/OsLoader.c [hunk 10]
```diff
@@ -154,8 +143,129 @@ ParseIasImage (
...
+  ZeroMem (File, sizeof (File));
+
+  DEBUG ((DEBUG_INFO, "CONTAINER size = 0x%x, image type = 0x%x, # of components = %d\n", LoadedImage->IasImage.Size, ContainerHdr->ImageType, ContainerHdr->Count));
+
+  // Enumerate all components
+  Index = 0;
+  ComponentName = 0;
+  while (TRUE) {
+    Status = GetNextAvailableComponent (ContainerHdr->Signature, (UINT32 *)&ComponentName);
+    if (EFI_ERROR(Status) || ((UINT32)ComponentName == SIGNATURE_32 ('_', 'S', 'G', '_')) ) {
```
Comment (myguitar) [inline PayloadPkg/OsLoader/OsLoader.c]: Don't we already have SIGNATURE_32 ('_', 'S', 'G', '_') definition? A single definition would be better.

File: PayloadPkg/OsLoader/OsLoader.c [hunk 11]
```diff
@@ -154,8 +143,129 @@ ParseIasImage (
...
+        // Use Load to decompress to a new buffer
+        Status = LoadComponent (ContainerHdr->Signature, (UINT32) ComponentName, (VOID **)&File[Index].Addr, &File[Index].Size);
+      }
+    }
+  }
+
+  Status = UnregisterContainer (ContainerHdr->Signature);
+  DEBUG ((DEBUG_INFO, "Unregister done - %r!\n", Status));
+
+  Status = UpdateLoadedImage (ContainerHdr->Count, File, LoadedImage, ContainerHdr->ImageType & 0xF);
```
Comment (myguitar) [inline PayloadPkg/OsLoader/OsLoader.c]: Need a comment about 0xF mask.

File: BuildLoader.py [hunk 12]
```diff
@@ -190,6 +190,7 @@ def __init__(self, *args, **kwargs):
 		self.ENABLE_GRUB_CONFIG    = 0
 		self.ENABLE_SMBIOS         = 0
 		self.ENABLE_LINUX_PAYLOAD  = 0
+		self.ENABLE_CONTAINER_BOOT = 1
```
Comment (myguitar) [inline BuildLoader.py]: Enabled by default?

File: PayloadPkg/OsLoader/LoadImage.c [hunk 13]
```diff
@@ -1,571 +1,579 @@
...
-    }
-
-    Status = GetFileSize (FileHandle, &ConfigFileSize);
-    if (EFI_ERROR (Status)) {
-      DEBUG ((DEBUG_INFO, "Get file '%s' size failed, Status = %r\n", (CHAR16 *)mConfigFileName[Index], Status));
-      CloseFile (FileHandle);
-      continue;
-    }
-
+/** @file
```
Comment (myguitar) [inline PayloadPkg/OsLoader/LoadImage.c]: It looks line ending changed. Can you keep previous line ending feed for better code review? Let's clean up line ending issue later in entire files.

File: BuildLoader.py (hunk 12 above)
Comment (stalamudupula) [reply BuildLoader.py]: Moving forward, the plan is to retire IAS, so making container image boot as default.

Comment (myguitar) [reply BuildLoader.py]: Got it. Thanks.

**Pull request merged**